process explorer/backend/ProcessExplorer/Processes/ProcessMonitor.cs

```diff
@@ -0,0 +1,375 @@
+                processInfoManager = ProcessInfoManagerFactory.SetProcessInfoGeneratorBasedOnOS(SendNew, SendDeleted, SendModified);
+        }
+
+        #endregion
+
+        #region Getters
+        public SynchronizedCollection<ProcessInfoDto>? GetProcesses()
+        {
+            ClearProcesses();
+            FillListWithRelatedProcesses();
+            return Data?.Processes;
+        }
+
+        internal IEnumerable<ProcessInfoDto>? GetProcessByName(string name)
+            => Data?.Processes.Where(p => p.ProcessName == name);
+
+        internal IEnumerable<ProcessInfoDto>? GetProcessByID(int id)
+            => Data?.Processes.Where(p => p.PID == id);
+        #endregion
+
+        #region Kill process
+        private async Task KillProcess(ProcessStartInfo info)
+        {
+            try
+            {
+                info.RedirectStandardError = true;
+                info.RedirectStandardInput = true;
+                info.RedirectStandardOutput = true;
+                info.UseShellExecute = false;
+                info.CreateNoWindow = true;
+
+                Process process = new Process();
+                process.StartInfo = info;
+                process.Start();
+                process.StandardOutput.ReadToEnd();
+                await process.WaitForExitAsync();
+            }
+            catch (Exception exception)
+            {
+                logger?.LogError(exception.Message);
+            }
+        }
+
+        public void KillProcessByName(string processName)
+        {
+            if (processInfoManager != null)
+                KillProcess(processInfoManager.KillProcessByName(processName)).Wait();
+        }
+
+        public void KillProcessById(int processId)
+        {
+            if (processInfoManager != null)
+                KillProcess(processInfoManager.KillProcessById(processId)).Wait();
+        }
+        #endregion
+
+        #region Delete terminated process from the list helper
+        private async Task<bool> TryDeleteMainProcesses(int pid)
+        {
+            try
+            {
+                var item = Data.Processes.Single(p => p.PID == pid);
+                if (item != default)
+                {
+                    Data.Processes.Remove(item);
+                    if (ProcessChangedPushingUrl != null)
+                        await ProcessStatusChanged(pid);
+                    return true;
+                }
+            }
+            catch (Exception exception)
+            {
+                logger?.LogError(exception.Message);
+                return false;
+            }
+            return false;
+        }
+
+        private async void TryDeleteChildProcesses(int pid)
+        {
+            try
+            {
+                lock (Data.Processes)
+                {
+                    foreach (var process in Data.Processes)
+                    {
+                        if (process.Children != default)
+                            lock (process.Children)
```

Comment:
I don't think we need a second lock here. There's a lock on a common object (Data.Processes) which prevents other threads from entering this block anyway.